### PR TITLE
Documentation and README gardening

### PR DESCRIPTION
**Pull request by zwaldowski**
Hi there, 2.0! (More or less)

#### What's in this pull request?

Address #70 and #84 through improvements minor changes to documentation, improvements to doc-set generation, and README reworking.

#### Testing

Not applicable.

#### API Changes

Not applicable.

### CODE REVIEW

**Comment by zwaldowski** [comment]
Be sure to check out the [rich diff](https://github.com/bignerdranch/Deferred/pull/94/files?short_path=04c6e90#diff-04c6e90faac2675aa89e2176d2eec7d8) view of the README! :smile:

**Comment by davidahouse** [comment]
@zwaldowski Looks great!

In the Combining Deferreds section, should this line be `joinedValues` instead of `all`? Or should we change the example to use `all` instead? Or should we have an example of both?

```
// `all` creates a new Deferred that is filled once all inputs are available.

...

let allDeferreds: Future<[Int]> = deferreds.joinedValues
```

**Comment by zwaldowski** [comment]
`joinedValues` replaces `all`. Good catch!

**Comment by davidahouse** [comment]
Found another cleanup item. The `Package.swift` file contains the wrong url. It should be:

```
.Package(url: "https://github.com/bignerdranch/Deferred.git", majorVersion: 1)
```

**Comment by zwaldowski** [comment]
Nope, this upcoming release is 2.0.

**Comment by davidahouse** [comment]
@zwaldowski OK, so it has TWO problems. :)

currently is:
```
 .Package(url: "https://github.com/bignerdranch/AtomicSwift.git", majorVersion: 1)
```

**Comment by zwaldowski** [comment]
**Oh** you were looking at the Package.swift.

No, the SPM manifest is correct. Deferred itself is a package that depends (for now) on the [AtomicSwift](https://github.com/bignerdranch/AtomicSwift/blob/master/Package.swift) pseudo-package.

**Comment by davidahouse** [comment]
@zwaldowski Ahh right, in the dependencies section. Sorry was scanning and missed the context. Carry on!

File: .jazzy.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# jazzy --swift-version 2.2 -m Deferred -g "https://github.com/bignerdranch/Deferred" -a "Big Nerd Ranch" -u "http://bignerdranch.github.io" --module-version=2.0.0 -r "http://bignerdranch.github.io/Deferred/"
```
Comment (jeremy-w) [inline .jazzy.yml]: For scripts, I prefer to use the long version of options where possible. The abbreviation when typing them out yourself of short options isn't needed, and the explanatory power of long options is wonderful.

If jazzy'll let you, I often spread args in a group across lines using continuation backslashes. (This happens a lot with gnarly `curl` commands.) Here's an example of what I mean from a Makefile I have lying around:

```make
run:
	browser-sync start \
		--server \
		--files "stylesheets/*.css, scripts/*.js, *.html"
```

Comment (jeremy-w) [inline .jazzy.yml]: On second thought, looks like this comment line was for copy-paste execution at the CLI. But it's out-of-date now we need so many options passed to jazzy. Nuke the line, or explain how to invoke it with this config file?

Comment (zwaldowski) [inline .jazzy.yml]: Ah! I didn't mean to leave this in. It should be sufficient to use the config file. Will note that instead.

File: .jazzy.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+author: "Big Nerd Ranch"
+author_url: "http://bignerdranch.github.io"
+copyright: "© 2014-2016 [Big Nerd Ranch](https://www.bignerdranch.com) under [MIT license](https://github.com/bignerdranch/Deferred/blob/master/LICENSE.txt)."
```
Comment (jeremy-w) [inline .jazzy.yml]: I'd make this "(C) DATE COMPANY. Licensed under the terms of the [MIT license]."

I also tend to use scheme-relative links like `//www.bignerdranch.com`, but that's not a big deal.

I'd be tempted to link the MIT license bit to the TLDR;Legal summary page: https://tldrlegal.com/license/mit-license#summary

Comment (zwaldowski) [inline .jazzy.yml]: Are scheme-relative links valid Markdown?

Comment (zwaldowski) [inline .jazzy.yml]: Doesn't seem like it http://meta.stackexchange.com/questions/121293/markdown-support-for-urls-without-a-protocol

Comment (zwaldowski) [inline .jazzy.yml]: Sorry, right. Clarification. The `copyright` stanza is included as Markdown in the footer of the docs.

File: .jazzy.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+custom_categories:
+  - name: Futures and Promises
+    children:
+    - Deferred
+    - FutureType
+    - PromiseType
+    - Timeout
+  - name: Future Mastery
```
Comment (jeremy-w) [inline .jazzy.yml]: This category stands out oddly amongst the other categories. I'm not sure what to rename it, or how Future contrasts to FutureType.

Comment (zwaldowski) [inline .jazzy.yml]: Yeah, I'm unsure as well. "Things That Are Futures But Are Not Deferred" seemed too wordy.

File: .jazzy.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    - CollectionType
+    - SequenceType
+    - IgnoringFuture
+  - name: Custom Execution
+    children:
+    - ExecutorType
+    - NSOperationQueue
+    - NSRunLoop
+    - CFRunLoop
+  - name: Data Locking
```
Comment (jeremy-w) [inline .jazzy.yml]: Suggest "Coordinating Access to Data".

File: .jazzy.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    children:
+    - LockProtected
+    - ReadWriteLock
+  - name: Locks
```
Comment (jeremy-w) [inline .jazzy.yml]: Suggest "Coordination Primitives".

File: .jazzy.yml (hunk 4 above)
Comment (zwaldowski) [inline .jazzy.yml]: :+1:

File: .jazzy.yml (hunk 5 above)
Comment (zwaldowski) [inline .jazzy.yml]: :OK:

File: .jazzy.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    - FutureType
+    - PromiseType
+    - Timeout
+  - name: Future Mastery
+    children:
+    - Future
+    - CollectionType
+    - SequenceType
+    - IgnoringFuture
+  - name: Custom Execution
```
Comment (jeremy-w) [inline .jazzy.yml]: Suggest "Controlling Where Callbacks Execute".

**Comment by jeremy-w** [comment]
Is there somewhere I can view the prepped docs online, so I can see how the categories look in context?

**Comment by zwaldowski** [comment]
I'll push it up to a bucket really quick

File: README.md [hunk 7]
```diff
@@ -1,82 +1,67 @@
...
-```
-
-You will also need to make sure you're opting into using frameworks:
-
-```ruby
-use_frameworks!
-```
-
-Then run `pod install`.
+This is an implementation of [OCaml's Deferred](https://ocaml.janestreet.com/ocaml-core/111.25.00/doc/async_kernel/#Deferred) for Swift. It lets you work with values that haven't been determined yet, like an array that's coming later (one day!) from a web service call.
```
Comment (jeremy-w) [inline README.md]: Let's reverse these two sentences:

> Deferred lets you work…call. It was originally inspired by [OCaml's Deferred library.](https://ocaml.janestreet.com/ocaml-core/111.25.00/doc/async_kernel/#Deferred)

File: README.md [hunk 8]
```diff
@@ -1,82 +1,67 @@
...
+
+### Vital Statistics
+
+|             | **Status**                                                                                                                                                  |
+|-------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------|
```
Comment (jeremy-w) [inline README.md]: This table is marvelously unreadable in monospace. :)

File: README.md [hunk 9]
```diff
@@ -131,23 +115,23 @@ This gets really messy with callbacks. But with `Deferred`, it's just:
...
     // names has been determined - use it!
     dataSource.array = names
     tableView.reloadData()
 }
 ```
 
-## Tasks
+## Basic Tasks
 
 ### Vending a Future
```
Comment (jeremy-w) [inline README.md]: Now that we have an actual `Future` type, we should probably be more careful about Future/Deferred verbiage. We probably do want to counsel them to vend a Future and not a Deferred, though.

File: README.md [hunk 10]
```diff
@@ -322,9 +321,154 @@ func AsynchronousFriends(forUser: jimbob) -> Deferred<[Friend]> {
...
-### Further Information
+## Mastering (the) `Future`
+
+Deferred is designed to scale with the fundamentals you see above. Large applications can be built using just `Deferred` and its `upon` and `fill` methods.
+
+### Read-Only Views
+
+It sometimes just doesn't make *sense* to be able to `fill` something; if you have a `Deferred` wrapping `UIApplication`'s push notification token, what does it mean if someone in your codebase calls `fill` on it?
+
+You may have noticed that anybody can call `upon` and `fill` on `Deferred` at any time; it's **read-write**. The former is fundamental, but the latter may be a liability as different pieces of code interact with each other. There's no concern to be had about thread safety, but instead about access: how do you limit illogical control flow, make it **read-only**?
```
Comment (jeremy-w) [inline README.md]: It's not clear to me what "former" and "latter" refer to here. Rewrite to avoid them.

File: README.md (hunk 8 above)
Comment (zwaldowski) [inline README.md]: Make the width _really_ big. :grimacing:

(Maybe I can use a URL shortener.)

File: README.md (hunk 10 above)
Comment (jeremy-w) [inline README.md]: I don't get the "illogical control flow" bit. What control flow? I think that's unnecessary though. We can replace this whole sentence with just, "How can we make it read-only?"

**Comment by zwaldowski** [comment]
https://bucket.waldowski.me/DeferredDocs20160412

File: README.md (hunk 9 above)
Comment (zwaldowski) [inline README.md]: I think we would be well served by

* "Future" (the idea)
* "the `Future` type" or "the `Future<Value>` type"
* "Deferred" (the framework)
* "the `Deferred` type" or "the `Deferred<Value>` type"

File: .jazzy.yml (hunk 2 above)
Comment (jeremy-w) [inline .jazzy.yml]: It does. The answer you wanted to read is:
http://meta.stackexchange.com/a/121321 The asker wasn't writing a valid
URL; the scheme-relative ones begin with a leading //, not directly with
the hostname.

I'm using it already on several GitHub READMEs, so. :) To see it in action,
check the raw view here:
https://github.com/bignerdranch/jws-coffeerun/blob/master/README.md

--
Jeremy W. Sherman <jeremy@bignerdranch.com>

2016-04-12 13:41 GMT-04:00 Zachary Waldowski <notifications@github.com>:

Comment (zwaldowski) [inline .jazzy.yml]: OK, sweet. I'll change it

File: README.md (hunk 8 above)
Comment (jeremy-w) [inline README.md]: I think using the footnote-style version would work, like:

[mumble][1]

  [1]: url goes here

That can be a string rather than a 1, and I recommend that in order to
avoid conflicts. :) You can stick the link-footnotes right under the table
so they stay by the text using them.

--
Jeremy W. Sherman <jeremy@bignerdranch.com>

2016-04-12 13:50 GMT-04:00 Zachary Waldowski <notifications@github.com>:

File: README.md (hunk 9 above)
Comment (jeremy-w) [inline README.md]: I like that. Maybe even go with "Vending a Future Value" to stick with the
"this is a value that hasn't quite arrived yet". Then we get future value
vs future type verbal distinction, even absent any fancy markup.

--
Jeremy W. Sherman <jeremy@bignerdranch.com>

2016-04-12 13:58 GMT-04:00 Zachary Waldowski <notifications@github.com>:

File: README.md (hunk 8 above)
Comment (zwaldowski) [inline README.md]: I always forget about those! Will play with it, thanks

File: .jazzy.yml (hunk 3 above)
Comment (jeremy-w) [inline .jazzy.yml]: Maybe "Controlling Who Can Fill" or "Access Control"? And shift `FutureType` in here alongside `Future`?

Comment (zwaldowski) [inline .jazzy.yml]: I'm not sure that approach is right for the framework at large, it just made most sense for the README.

Comment (zwaldowski) [inline .jazzy.yml]: Eh. Maybe you're right.

File: .jazzy.yml (hunk 2 above)
Comment (zwaldowski) [inline .jazzy.yml]: This conveniently breaks when I'm viewing locally (`file://www.bignerdranch.com`), but so be it

Comment (jeremy-w) [inline .jazzy.yml]: Ah, fun. Try viewing on `http://localhost:PORT` via `python -mSimpleHTTPServer` or [devd](https://github.com/cortesi/devd) perhaps?

File: README.md [hunk 11]
```diff
@@ -329,7 +329,7 @@ Deferred is designed to scale with the fundamentals you see above. Large applica
 
 It sometimes just doesn't make *sense* to be able to `fill` something; if you have a `Deferred` wrapping `UIApplication`'s push notification token, what does it mean if someone in your codebase calls `fill` on it?
 
-You may have noticed that anybody can call `upon` and `fill` on `Deferred` at any time; it's **read-write**. The former is fundamental, but the latter may be a liability as different pieces of code interact with each other. There's no concern to be had about thread safety, but instead about access: how do you limit illogical control flow, make it **read-only**?
+You may have noticed that anybody can call `upon` on a `Deferred` type; this is fundamental. But the same is true of `fill`, and this may be a liability as different pieces of code interact with each other. How can we make it **read-only**?
```
Comment (jeremy-w) [inline README.md]: This is much clearer. :grapes:

File: Sources/ExecutorType.swift [hunk 12]
```diff
@@ -55,6 +83,12 @@ struct QueueExecutor: ExecutorType {
 
 }
 
+/// An operation queue manages a number of operation objects, making higher-
```
Comment (jeremy-w) [inline Sources/ExecutorType.swift]: Does this correctly hyphenate as `higher-level` in Xcode and Jazzy, or does it end up rendering as `higher- level` with a space in the middle due to the linebreak?

File: README.md [hunk 13]
```diff
@@ -322,9 +326,154 @@ func AsynchronousFriends(forUser: jimbob) -> Deferred<[Friend]> {
...
+
+It sometimes just doesn't make *sense* to be able to `fill` something; if you have a `Deferred` wrapping `UIApplication`'s push notification token, what does it mean if someone in your codebase calls `fill` on it?
+
+You may have noticed that anybody can call `upon` on a `Deferred` type; this is fundamental. But the same is true of `fill`, and this may be a liability as different pieces of code interact with each other. How can we make it **read-only**?
+
+For this reason, Deferred is split into `FutureType` and `PromiseType`, both protocols the `Deferred` type conforms to. You can think of these as the "reading" and "writing" sides of a deferred value; a future can only be `upon`ed, and a promise can only be `fill`ed.
+
+Deferred also provides the `Future` type, a wrapper for anything that's a `FutureType` much like the Swift standard library's `Any` types. You can use it protectively to make a `Deferred` read-only. Reconsider the example from above:
+
+```
```
Comment (jeremy-w) [inline README.md]: Should we ensure all code blocks are have the language marked, as:

    ```swift

and not just:

    ```

?

File: Sources/ExecutorType.swift (hunk 12 above)
Comment (zwaldowski) [inline Sources/ExecutorType.swift]: Hilariously, neither. The "NSOperationQueue Extension Reference" does not include this doc at all.

Moved, it is the dumb `higher- level`. I'll fix it.

File: README.md (hunk 13 above)
Comment (zwaldowski) [inline README.md]: Fenced!
